pkg/lib/cockpit/dialog.tsx

```diff
@@ -0,0 +1,1543 @@
+         comes to a conclusion, and the path is still relevant to the
+         dialog, but this promise is no longer the current
+         promise. Its result will be ignored, as it should.
+     */
+
+    #validation_state_is_current(state: DialogValidationState, prom?: Promise<void>): boolean {
+        return (
+            (!prom || Object.is(state.promise, prom)) &&
+                this.#is_current_validation_round_id(state.round_id)
+        );
+    }
+
+    /* _validate_value_async puts this all together.
+     */
+
+    _validate_value_async(path: string, val: unknown, debounce: number, func: () => Promise<string | undefined>): void {
+        const state = this.#get_validation_state(path);
+        if (!this.#probe_validation_state_cache(state, val)) {
+            debug("async validate start debounce", state.path, val);
+            this.#set_validation_state_timeout(
+                state,
+                val,
+                debounce,
+                () => {
+                    debug("async validate start promise", state.path, val);
+                    const prom =
+                        func()
+                                .catch(
+                                    ex => {
+                                        console.error(ex);
+                                        return undefined;
+                                    }
+                                )
+                                .then(
+                                    result => {
+                                        if (this.#validation_state_is_current(state, prom)) {
+                                            debug("async validate done", state.path, result);
+                                            this.#set_validation_state_result(state, val, result);
+                                        } else {
+                                            debug("promise outdated", state.path);
+                                        }
+                                    }
+                                );
+                    this.#set_validation_state_promise(state, val, prom);
+                }
+            );
+        }
+    }
+
+    /* Since the DialogValidationState for a path is so important, it
+       is also important to keep them firmly associated with each
+       other when the path of a value changes.
+
+       A path might change when there are arrays involved, and
+       elements get new indices without actually changing identity.
+     */
+
+    _rename_validation_state(path: string, from: number, to: number) {
+        const from_path = path + "." + String(from);
+        const to_path = path + "." + String(to);
+        if (from_path in this.#validation_state) {
+            debug("rename", from_path, to_path);
+            this.#validation_state[to_path] = this.#validation_state[from_path];
+            this.#validation_state[to_path].path = to_path;
+            delete this.#validation_state[from_path];
```

Comment:
These 5 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22700-2dd5f8c0-20251219-103022-fedora-43-devel-other/Coverage/lcov/github-pr.diff.gcov.html)